frame/oracle/src/tests.rs

```diff
@@ -20,205 +20,462 @@ use sp_runtime::{
-		let account_3 = get_account_3();
-		let account_4 = get_account_4();
-		let account_5 = get_account_5();
 
-		assert_ok!(Oracle::set_signer(Origin::signed(account_2), account_1));
-		assert_eq!(Oracle::controller_to_signer(account_2), Some(account_1));
-		assert_eq!(Oracle::signer_to_controller(account_1), Some(account_2));
+mod add_asset_and_info {
+    use super::*;
+
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(10_000))]
+
+        #[test]
+        fn normal_asset_and_info_assert(
+            asset_id in asset_id(),
+            asset_info in asset_info(),
+        ) {
+            new_test_ext().execute_with(|| {
+                let root_account = get_root_account();
+
+                prop_assert_ok!(Oracle::add_asset_and_info(
+                    Origin::signed(root_account),
+                    asset_id,
+                    asset_info.threshold,
+                    asset_info.min_answers,
+                    asset_info.max_answers,
+                    asset_info.block_interval,
+                    asset_info.reward,
+                    asset_info.slash,
+                ));
+
+                Ok(())
+            })?;
+        }
+
+        #[test]
+        fn asset_count_should_not_increase_when_updating_asset_info(
+            asset_id in asset_id(),
+            asset_info_1 in asset_info(),
+            asset_info_2 in asset_info(),
+        ) {
+            new_test_ext().execute_with(|| {
+                let root_account = get_root_account();
+
+                prop_assert_ok!(Oracle::add_asset_and_info(
+                    Origin::signed(root_account),
+                    asset_id,
+                    asset_info_1.threshold,
+                    asset_info_1.min_answers,
+                    asset_info_1.max_answers,
+                    asset_info_1.block_interval,
+                    asset_info_1.reward,
+                    asset_info_1.slash,
+                ));
+
+                // does not increment asset_count because we have info for the same asset_id
+                prop_assert_ok!(Oracle::add_asset_and_info(
+                    Origin::signed(root_account),
+                    asset_id,
+                    asset_info_2.threshold,
+                    asset_info_2.min_answers,
+                    asset_info_2.max_answers,
+                    asset_info_2.block_interval,
+                    asset_info_2.reward,
+                    asset_info_2.slash,
+                ));
+                prop_assert_eq!(Oracle::assets_count(), 1);
+
+                Ok(())
+            })?;
+        }
+
+        #[test]
+        fn fails_when_non_root_account(
+            asset_id in asset_id(),
+            asset_info in asset_info(),
+            account_id in account_id(),
+        ) {
+            // very small chance of happening, but for correctness' sake ;)
+            prop_assume!(account_id != get_root_account());
+
+            new_test_ext().execute_with(|| {
```

Comment:
no sure if should be prop test so. it is easy to spot if somebody changes root to non root in review. probably we will have that non root in future. and non prop test is enough i think. prop test cost - time to maintain and time to run in ci